### PR TITLE
Wallet file standard

### PR DESCRIPTION
**Pull request by Thykof**

### CODE REVIEW

File: wallet/file-format.md [hunk 1]
```diff
@@ -2,10 +2,91 @@
...
-## Abstract
+**Authors:** N.Seva, G. Libert
 
-## Motivation
+**Status:** Draft
+
+**Version:** 0.1
+
+## Introduction
+The Wallet file format standard provides a specification for protection and serialization of accounts within a wallet. This standard doesn't specify the address object nor the cryptography used by the massa blockain to sign operation.
```
Comment (mazmassa) [inline wallet/file-format.md:12]:
```suggestion
The Wallet file standard provides a specification for protection and serialization of accounts within a wallet. This standard doesn't specify the address object nor the cryptography used by the Massa blockain to sign operation.
```

File: wallet/file-format.md [hunk 2]
```diff
@@ -2,10 +2,115 @@
...
+
+The encryption and decryption of the private key must be performed correctly to ensure the confidentiality and integrity of the account. If the encryption algorithm or key derivation function is weak, the private key may be vulnerable to attacks. It is therefore crucial to use strong and proven encryption algorithms and key derivation functions.
+
+### Salt Generation
+
+The generation of random and unpredictable values for the salt is critical in protecting the private key. If the values are not truly random or can be easily guessed, the encryption of the private key may be compromised. It is therefore essential to use a robust and reliable method.
+
+### Password Strength
+
+The strength of the user's password is crucial to the security of the wallet. If the password is weak or easily guessable, the private key's confidentiality may be compromised. Therefore, it is essential to educate users on creating strong passwords and implementing password policies that enforce minimum complexity requirements.
```
Comment (peterjah) [inline wallet/file-format.md:109]: Altough i agree with this is general, I don't really understand the presence of this section, as there is no mention of password in this spec.

Comment (Thykof) [reply wallet/file-format.md:109]: see line 43 in PBKDF2 section.

**Comment by dr-chain** [comment]
[Inputs from the Security Team]
After discussing with @gregLibert, we can go for a non-HD wallet standard for at least version 1. HD Wallet Standard can be added in another version as it involves multiple implementations before it can go live.

I am adding other comments here:
1. We are using AES-GCM for encryption/decryption which also uses Authentication. So following this, the TAG from AES GCM should also be stored in the file otherwise we will never know if someone provided the right password as even for the wrong password, there is a possibility we decrypt the private key to a wrong value. (since it is just a random sequence of bytes).

Since we do not assume that private and public keys are linked, the TAG is even more important.
So, I suggest adding TAG as well to the standard and a subroutine in the code too to check the authenticity of the password using this TAG. 

@gregLibert Mentioned that the TAG is already prepended to the Encrypted Private Key. If that is the case, we can either specify it or separate the TAG for compatibility and clarity with legacy systems. (He is investigating it currently)

2. Any specific reason to call IV (Initialization Vector) a nonce?
3. Can we call the Private key an Encrypted Private Key or something similar for clarity?

**Comment by gregLibert** [comment]
> 2\. Any specific reason to call IV (Initialization Vector) a nonce?

Because it's named like that in the specification itself and inthe implementation in Golang.

**Comment by gregLibert** [comment]
> 3\. Can we call the Private key an Encrypted Private Key or something similar for clarity?

Sounds good to me.

**Comment by gregLibert** [comment]
> 4\. Increase the iterations to 600k following the recommendations of [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2)

We need to check the consequences of moving from 10 000 to 600 000 before enforcing the change. This needs to be tested on both wallet provider: thyra and node-client.

**Comment by dr-chain** [comment]
"Tag itself is part of the [specification](https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38d.pdf), only the size is unclear as 7 values are possible:"

Yes, it is but do we include it in the encrypted output or as a separate thing since the size can change w.r.t. libraries?

**Comment by gregLibert** [comment]
> Yes, it is but do we include it in the encrypted output or as a separate thing since the size can change w.r.t. libraries?

Understood. Right now the tag is prepend in the [go implementation](https://go.dev/src/crypto/cipher/gcm.go#L191). If it's the same thing for the rust one, I propose to specify it like that.

If it's not, I propose to use another field to store it.

**Comment by gregLibert** [comment]
> > > Yes, it is but do we include it in the encrypted output or as a separate thing since the size can change w.r.t. libraries?
> > 
> > 
> > Understood. Right now the tag is prepend in the [go implementation](https://go.dev/src/crypto/cipher/gcm.go#L191). If it's the same thing for the rust one, I propose to specify it like that.
> > If it's not, I propose to use another field to store it.
> 
> Ok, that is great. In case the TAG is always prepended, then we just need to mention the default size we expect in the standard, explicitly. :)

From my understanding of https://docs.rs/aead/latest/aead/trait.Aead.html#tymethod.encrypt:

image

rust is also using a postfix tag (I corrected my initial message), so I will only add the size in the specification and explain how it should be serailized and I will let the core team confirm all that when they will have some time to check that :-)

File: wallet/file-format.md [hunk 3]
```diff
@@ -2,10 +2,121 @@
...
+### Cryptography
+
+#### PBKDF2
+
+To protect the private key, a symmetric key is derived from the user password using the PBKDF2 algorithm, as defined in IETF [RFC 2898](https://www.ietf.org/rfc/rfc2898.txt).
+
+Specifically, the PBKDF2 arguments defined in section 5.2 of the aforementioned standard must follow the followings:
+
+- 16-byte salt,
+- 10,000 iterations, and a
```
Comment (Thykof) [inline wallet/file-format.md:48]: To keep track of the comment:

> 
> 4. Increase the iterations to 600k following the recommendations of [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2)
> 
> We need to check the consequences of moving from 10 000 to 600 000 before enforcing the change. This needs to be tested on both wallet provider: thyra and node-client.

**Comment by Thykof** [comment]
The tag is also at the end of the cipher text in node js implementation: https://github.com/nodejs/node/blob/main/src/crypto/crypto_aes.cc#L156

File: wallet/file-format.md [hunk 4]
```diff
@@ -2,10 +2,124 @@
...
+
+Specifically, the PBKDF2 arguments defined in section 5.2 of the aforementioned standard must follow the followings:
+
+- 16-byte salt,
+- 600,000 iterations, and a
+- derived key length of 32 bytes.
+
+The hash function utilized in this process is SHA-256, as specified in the NIST [FIPS 180-4](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf) document.
+
+These values align with the recommendations set forth in the NIST [Special Publication 800-132](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-132.pdf).
```
Comment (gregLibert) [inline wallet/file-format.md:53]: The 600,000 iterations isnot from the Nist but from the OWASP.

**Review by gregLibert — CHANGES_REQUESTED** [review]

**Review by dr-chain — CHANGES_REQUESTED** [review]
We will need to explain this as elaborated by @gregLibert 
" I will only add the size in the specification and explain how it should be serialized and I will let the core team confirm all that when they will have some time to check that :-)"

**Review by gregLibert — APPROVED** [review]

**Review by gregLibert — CHANGES_REQUESTED** [review]

**Review by BatiGencho — APPROVED** [review]

**Review by dr-chain — CHANGES_REQUESTED** [review]
- Do you know what is the nonce size we are currently proposing in the standard? I think it is 12 bytes (96 bits) but I do not see that written in the standards file. Maybe, can we add this information?

- We also need to add that we use input password converted to bytes using the utf-8 encoding, so that everything is clear. I think this bit of information is missing 🙂

**Review by dr-chain — APPROVED** [review]

**Comment by dr-chain** [comment]
@Thykof @gregLibert I made few changes and added few things. Also changed the author list and the status.
Please have a look.

**Pull request merged**